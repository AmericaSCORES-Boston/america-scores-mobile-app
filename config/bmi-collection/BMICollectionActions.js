export const START_BMI_COLLECTION = "START_BMI_COLLECTION";
export const END_BMI_COLLECTION = "END_BMI_COLLECTION";
export const SAVE_STUDENT = "SAVE_STUDENT";
export const SAVE_STUDENT_SUCCESS = "SAVE_STUDENT_SUCCESS";
export const SAVE_STUDENT_FAILURE = "SAVE_STUDENT_FAILURE";

export function startBMICollection(studentsToFilter) {
    return {
        type: START_BMI_COLLECTION,
        students: studentsToFilter
    }
}

export function endBMICollection(studentsToFilter) {
    return {
        type: END_BMI_COLLECTION,
        students: studentsToFilter
    }
}

export function saveStudent(student) {
    return {
        type: SAVE_STUDENT,
        student: student
    }
}

export function onSaveStudentSuccess(student) {
    return {
        type: SAVE_STUDENT_SUCCESS,
        student: student

    }
}

export function onSaveStudentFailure(student) {
    return {
        type: SAVE_STUDENT_FAILURE,
        student: student
    }
}