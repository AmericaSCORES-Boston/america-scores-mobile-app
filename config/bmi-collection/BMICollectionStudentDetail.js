const React = require('react');

module.exports = class BMICollectionStudentDetail extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <div
                className="bmi-collection-student-detail">
                <span>{this.props.student.id}</span>
                <span>{this.props.student.name}</span>
                <span>{this.props.student.dob}</span>
                <span>{this.props.student.height}</span>
                <span>{this.props.student.weight}</span>
            </div>
        );
    }
};