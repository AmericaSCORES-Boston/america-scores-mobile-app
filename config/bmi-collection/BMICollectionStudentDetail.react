import React from 'react';

export default class StudentDetail extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <div
                className="student-detail">
                <span>{this.props.student.id}</span>
                <span>{this.props.student.name}</span>
                <span>{this.props.student.dob}</span>
                <span>{this.props.student.height}</span>
                <span>{this.props.student.weight}</span>
            </div>
        );
    }

}