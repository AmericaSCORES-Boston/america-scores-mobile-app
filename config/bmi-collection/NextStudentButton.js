const React = require('react');

module.exports = class NextStudentButton extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <button className="next-student">Save</button>
        );
    }

};