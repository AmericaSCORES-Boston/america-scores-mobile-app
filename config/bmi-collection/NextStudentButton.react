import React from 'react';

export default class NextStudentButton extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <button className="next-student">Save</button>
        );
    }

}