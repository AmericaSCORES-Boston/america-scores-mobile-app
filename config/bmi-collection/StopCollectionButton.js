const React = require('react');

module.exports = class StopCollectionButton extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <button className="stop-collection">Stop</button>
        );
    }

};