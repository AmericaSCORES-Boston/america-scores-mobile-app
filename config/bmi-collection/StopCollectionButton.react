import React from 'react';

export default class StopCollectionButton extends React.Component {

    constructor() {
        super();
    }

    render() {
        return (
            <button className="stop-collection">Stop</button>
        );
    }

}