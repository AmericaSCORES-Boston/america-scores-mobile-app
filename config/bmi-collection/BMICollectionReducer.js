import * as actions from 'BMICollectionActions';

export const initialState = {
    currentStudent: "",
    students: []
};

// A Mocked implementation of the BMICollectionReducer.
export default function BMICollectionReducer(state = initialState, action) {
    switch (action.type) {
        case actions.START_BMI_COLLECTION:
            return state;
        case actions.END_BMI_COLLECTION:
            return state;
        case actions.SAVE_STUDENT:
            return state;
        case actions.SAVE_STUDENT_SUCCESS:
            return state;
        case actions.SAVE_STUDENT_FAILURE:
            return state;
        default:
            return state;
    }
}
