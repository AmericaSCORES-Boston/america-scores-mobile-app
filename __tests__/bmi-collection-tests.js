/**
* ********************************************
*   PAGE FLOW
* ********************************************
* 1) User begins BMI Collection
* 2) Student detail view is shown (students with information should *probably* not be shown)
* 3) User decides:
*    a) User enters info about student (repeat from step 3)
*    b) Clicks 'Save' (repeat from step 2)
*    c) Clicks 'Stop'
*    d) Clicks 'Previous Student' (repeat from step 2) (optional)
*
* ********************************************
*   COMPONENTS
* ********************************************
* -- StudentDetailsContainer (manages state)
*    -- StudentDetail (contains stats/info)
*    -- 'Save' Button (save this student and show next one)
*    -- 'Stop' Button
*    -- 'Previous Student' Button (optional)
*
* ********************************************
*   ACTIONS
* ********************************************
*  -- Start Collection (pass in current roster, filter students who don't have a height or weight entered)
*  -- End Collection
*
* -- Async:
*   -- Save Student:
*      -- Request to save student
*      -- Successful save of student
*      -- Failure to save student
*
* ********************************************
*   REDUCERS
* ********************************************
* -- BMICollection State (updates current student, and maintains list of students missing height or weight)
* -- StudentDetail State
*      -- Request to save student - show 'Saving data for [student].'
*      -- Success on saving student - show '[Student]'s data has been saved.' Update navigation
*      -- Failure to save student - show feedback and student with saved input
*
*/

'use strict';

import React from 'react';
import renderer from 'react-test-renderer';
import BMICollectionStudentDetail from '../config/bmi-collection/BMICollectionStudentDetail.js';
import NextStudentButton from '../config/bmi-collection/NextStudentButton.js';
import StopCollectionButton from '../config/bmi-collection/StopCollectionButton.js';
import * as MockStudents from '../config/bmi-collection/data/MockStudents.js';

/**
* *************************************************
*   COMPONENT TESTS
* *************************************************
*/

describe('component tests', () => {
  it('renders the student1 detail view correctly', () => {
    const tree = renderer.create(
      <BMICollectionStudentDetail student={MockStudents.student1}></BMICollectionStudentDetail>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders "Save" button correctly', () => {
    const tree = renderer.create(
      <NextStudentButton></NextStudentButton>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders "Stop" button correctly', () => {
    const tree = renderer.create(
      <StopCollectionButton></StopCollectionButton>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});



/**
* *************************************************
*   ACTION CREATOR TESTS
* *************************************************
*/
import * as actions from '../config/bmi-collection/BMICollectionActions';

describe('action creator tests', () => {
  it('creates a "START_BMI_COLLECTION" action', () => {
    expect(actions.startBMICollection(MockStudents.students)).toMatchSnapshot();
  });

  it('creates an "END_BMI_COLLECTION" action', () => {
    expect(actions.endBMICollection(MockStudents.students)).toMatchSnapshot();
  });

  // Async action tests may need to be reworked since they are more complicated,
  // but this is the basic idea.

  it('creates a "SAVE_STUDENT" action', () => {
    expect(actions.saveStudent(MockStudents.student1)).toMatchSnapshot();
  });

  it('creates a "SAVE_STUDENT_SUCCESS" action', () => {
    expect(actions.saveStudent(MockStudents.student1)).toMatchSnapshot();
  });

  it('creates a "SAVE_STUDENT_FAILURE" action', () => {
    expect(actions.saveStudent(MockStudents.badStudent)).toMatchSnapshot();
  });
});

/**
* *************************************************
*   REDUCER TESTS
* *************************************************
*/
import {BMICollectionReducer as reducer} from '../config/bmi-collection/BMICollectionReducer';

describe('reducer tests', () => {

  it('returns the same state on an unhandled action', () => {
    expect(reducer(undefined, {type: '_NULL'})).toMatchSnapshot();
  });

  it('starts BMI collection with a filtered list', () => {
    const action = actions.startBMICollection(MockStudents.students);
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  it('starts BMI collection', () => {
    const action = actions.startBMICollection(MockStudents.studentsWithoutHeightAndWeight);
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  it('ends BMI collection', () => {
    const action = actions.endBMICollection(MockStudents.studentsWithoutHeightAndWeight);
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  it('does not start BMI collection because there are no students created for this program', () => {
    const action = actions.startBMICollection([]);
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  it('does not start BMI collection because height and weight data has already been collected for all students', () => {
    const action = actions.startBMICollection(MockStudents.studentsWithHeightAndWeight);
    expect(reducer(undefined, action)).toMatchSnapshot();
  });

  it('makes a request to save the data entered for a student', () => {
    const action = actions.saveStudent(MockStudents.student1);
    expect(reducer(MockStudents.studentsWithoutHeightAndWeight, action)).toMatchSnapshot();
  });

  it('saves the student data successfully', () => {
    const action = actions.onSaveStudentSuccess(MockStudents.student1);
    expect(reducer(MockStudents.studentsWithoutHeightAndWeight, action)).toMatchSnapshot();
  });

  it('fails to save the student data', () => {
    const action = actions.onSaveStudentFailure(MockStudents.badStudent);
    expect(reducer(MockStudents.studentsWithoutHeightAndWeight, action)).toMatchSnapshot();
  });
});
