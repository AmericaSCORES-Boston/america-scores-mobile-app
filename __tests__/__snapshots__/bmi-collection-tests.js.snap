exports[`action creator tests creates a "SAVE_STUDENT" action 1`] = `
Object {
  "student": Object {
    "dob": "03-29-2002",
    "height": 15,
    "id": 1,
    "name": "John Doe",
    "weight": null,
  },
  "type": "SAVE_STUDENT",
}
`;

exports[`action creator tests creates a "SAVE_STUDENT_FAILURE" action 1`] = `
Object {
  "student": Object {
    "dob": "01-01-1900",
    "height": "tall",
    "id": 4,
    "living": false,
    "name": "Grim Reaper",
    "weight": null,
  },
  "type": "SAVE_STUDENT",
}
`;

exports[`action creator tests creates a "SAVE_STUDENT_SUCCESS" action 1`] = `
Object {
  "student": Object {
    "dob": "03-29-2002",
    "height": 15,
    "id": 1,
    "name": "John Doe",
    "weight": null,
  },
  "type": "SAVE_STUDENT",
}
`;

exports[`action creator tests creates a "START_BMI_COLLECTION" action 1`] = `
Object {
  "students": Array [
    Object {
      "dob": "03-29-2002",
      "height": 15,
      "id": 1,
      "name": "John Doe",
      "weight": null,
    },
    Object {
      "dob": "08-12-1996",
      "height": 10,
      "id": 2,
      "name": "Jeremy Parker",
      "weight": 50,
    },
    Object {
      "dob": "02-05-1997",
      "height": null,
      "id": 3,
      "name": "Sally Smith",
      "weight": null,
    },
  ],
  "type": "START_BMI_COLLECTION",
}
`;

exports[`action creator tests creates an "END_BMI_COLLECTION" action 1`] = `
Object {
  "students": Array [
    Object {
      "dob": "03-29-2002",
      "height": 15,
      "id": 1,
      "name": "John Doe",
      "weight": null,
    },
    Object {
      "dob": "08-12-1996",
      "height": 10,
      "id": 2,
      "name": "Jeremy Parker",
      "weight": 50,
    },
    Object {
      "dob": "02-05-1997",
      "height": null,
      "id": 3,
      "name": "Sally Smith",
      "weight": null,
    },
  ],
  "type": "END_BMI_COLLECTION",
}
`;

exports[`component tests renders "Save" button correctly 1`] = `
<button
  className="next-student">
  Save
</button>
`;

exports[`component tests renders "Stop" button correctly 1`] = `
<button
  className="stop-collection">
  Stop
</button>
`;

exports[`component tests renders the student1 detail view correctly 1`] = `
<div
  className="bmi-collection-student-detail">
  <span>
    1
  </span>
  <span>
    John Doe
  </span>
  <span>
    03-29-2002
  </span>
  <span>
    15
  </span>
  <span />
</div>
`;

exports[`reducer tests does not start BMI collection because height and weight data has already been collected for all students 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;

exports[`reducer tests does not start BMI collection because there are no students created for this program 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;

exports[`reducer tests ends BMI collection 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;

exports[`reducer tests fails to save the student data 1`] = `
Array [
  Object {
    "dob": "03-29-2002",
    "height": 15,
    "id": 1,
    "name": "John Doe",
    "weight": null,
  },
  Object {
    "dob": "02-05-1997",
    "height": null,
    "id": 3,
    "name": "Sally Smith",
    "weight": null,
  },
]
`;

exports[`reducer tests makes a request to save the data entered for a student 1`] = `
Array [
  Object {
    "dob": "03-29-2002",
    "height": 15,
    "id": 1,
    "name": "John Doe",
    "weight": null,
  },
  Object {
    "dob": "02-05-1997",
    "height": null,
    "id": 3,
    "name": "Sally Smith",
    "weight": null,
  },
]
`;

exports[`reducer tests returns the same state on an unhandled action 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;

exports[`reducer tests saves the student data successfully 1`] = `
Array [
  Object {
    "dob": "03-29-2002",
    "height": 15,
    "id": 1,
    "name": "John Doe",
    "weight": null,
  },
  Object {
    "dob": "02-05-1997",
    "height": null,
    "id": 3,
    "name": "Sally Smith",
    "weight": null,
  },
]
`;

exports[`reducer tests starts BMI collection 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;

exports[`reducer tests starts BMI collection with a filtered list 1`] = `
Object {
  "currentStudent": "",
  "students": Array [],
}
`;
